CIP-0031/README.md

```diff
@@ -0,0 +1,210 @@
+Comments-Summary: No comments
+Comments-URI: 
+Status: Draft
+Type: Standards Track
+Created: 2021-11-29
+License: CC-BY-4.0
+---
+
+# Reference inputs
+
+## Abstract
+
+We introduce a new kind of input, a _reference_ input, which allows looking at an output without spending it.
+This will facilitate access to information stored on the blockchain without the churn associated with spending and recreating UTXOs.
+
+## Motivation
+
+Datums in transaction outputs provide a way to store and access information on the blockchain.
+However, they are quite constrained in a number of ways.
+Most notably, in order to access the information which is contained in them, you have to _spend_ the output that the datum is attached to.
+
+This has a number of undesirable features:
+1. Even if the output is recreated after being spent, it is still a _new_ output. Any other user who wishes to look at the data cannot spend the old output (which is gone), but must rather spend the new output (which they will not know about until the next block). In practice this throttles some applications to one "operation" per block.
+2. Looking at the _information_ in the datum requires _spending_ the output, which means that care must be taken over the distribution of any funds in the output (possibly requiring scripts), and spending conditions must be met. This is overly stringent, inconvenient, and expensive.
+
+We would like to have a mechanism for accessing the information in datums that avoids these issues.
+
+### Use cases
+
+Here are some cases where we expect this to be helpful.
+
+1. Inspecting the state (datum, or locked value) of an on-chain application without having to consume the output, e.g. checking the current state of a stablecoin state machine.
+2. On-chain data providers that store data in outputs, to be referenced by other scripts.
+3. (With inline datums) Creating a single UTXO with data to be used in multiple subsequent transactions, but only paying the cost for submitting it once.
+
+### Enabling further improvements
+
+This proposal was designed in tandem with two further proposals which make use of reference scripts, CIP-32 (Inline datums) and CIP-33 (Reference scripts).
+A full assessment of the worth of this proposal should take into account the fact that it is an enabler for these other proposals.
+
+## Specification
+
+### Reference inputs
+
+We introduce a new kind of transaction input, a reference input.
+Transaction authors specify inputs as either normal (spending) inputs or reference inputs.[^1]
+
+A reference input is a transaction input, which is linked to a particular transaction output as normal, except that it _references_ the output rather than _spending_ it. That is:
+
+- A referenced output must exist in the UTXO set.
+- Any value on a referenced output is _not_ considered when balancing the transaction.
+- The spending conditions on referenced outputs are _not_ checked, nor are the witnesses required to be present. 
```

Comment:
You may want to clarify that the spending conditions are checked if the referenced output is spent? I assume this is the case, since not doing so constitutes a DOS attack.

Reply:
s/referenced output/reference input/ would be better IMO

Reply:
indeed

Reply:
I can make it explicit. Since there are no changes to outputs, there is no change to the behaviour of spending an output.